### ISSUE TITLE
[Proposal]: debounce fix

### ISSUE THREAD

**Issue by bigopon** [aurelia/templating-resources#319]
Currently debounce is working incorrectly because of [this line in binding-expression](https://github.com/aurelia/binding/blob/master/src/binding-expression.js#L68)

```js
// line 68
if (newValue !== this.sourceExpression.evaluate(this.source, this.lookupFunctions)) {
  this.updateSource(newValue);
}
```

Reverting the value quickly will cause the new `updateSource` to not be called.

Proposal fix: Debounce binding behavior will alter the binding even further, when `methodToDebounce` is `updateSource`:

  * Modify `call`: check if it's `targetContext`, we call `updateSource` directly and let the diff happen there. This works but is bad because `targetContext` is private constant of aurelia binding.
  * Modify `updateSource` with another version that has value diffing.
  * [Demo of fix](https://gist.run/?id=85c510c2c33388258fa08bf1f41f16b5) for aurelia/binding#625 & ~~aurelia/binding#485~~

@jdanyow @EisenbergEffect @salchi

**Comment by jdanyow** [comment]
@bigopon I think you're on the right track here. I'm experimenting with putting the debounce on the `call` method all the time: https://gist.run/?id=74fb883f778d60583ab7cb6806d43215

**Comment by bigopon** [comment]
It looks like I didn't take enough steps back. Is there any gotcha that holds you from releasing it ?

**Comment by jdanyow** [comment]
I'm pretty happy with the [prototype](https://gist.run/?id=74fb883f778d60583ab7cb6806d43215) now, I really like your idea to debounce the `call` method instead of `updateTarget` / `updateSource`. There's a few things that would need to be done before we could release this:

1. It depends on https://github.com/aurelia/binding/commit/970d56db9a73dff516cf51a44a645b0a3e39efe9
2. `& throttle` needs a similar refactoring
2. My original tests for this are disgusting and slow. I should have used [jasmine.clock()](https://jasmine.github.io/2.0/introduction.html#section-Mocking_the_JavaScript_Timeout_Functions)

Let me know if you want to continue on this :smile: :pray:

**Comment by bigopon** [comment]
What do you think about introducing a binding manager, as sometimes we want to flush all currently debounced binding updates.
Use case: When user input first name, we want to start validating after 1s but then user hit submit, we should flush the debounce immediately and start validating it with the new value instead of the old one.

Is this fix a good chance to deal with that ?

And sure, i'll fix the disgusting tests 😷

**Comment by jdanyow** [comment]
re: binding manager- this kind of api could be added to the [`BindingEngine`](https://github.com/aurelia/binding/blob/77ede7839e20d1acbb8ad997f0852d99ac9043d8/src/binding-engine.js#L13). Would need flush the dirty checker, then the task queue. Sounds like a good change for a separate PR?

re: debounce on input, immediate on blur.... that should be the default behavior for throttle and debounce. Do you agree? I'd be in favor of adding support for blur in this PR.

**Comment by bigopon** [comment]
Ok lets go for debounce with event then but I'm not sure if the default behavior should be in now as it causes breaking change.

**Comment by bigopon** [comment]
@jdanyow

* I think the feature immediate on blur, debounce on input better fits a separate PR. (not as simple as I thought)
* For `debounce`, it's perfect for jasmine clock as it 100% timeout. I finished this and test for it.
* For `throttle`, it's quite challenging if we want to install the clock, as it's based on `+ new Date()`. I think we need to keep your old test. But i'm getting error for the test when refactoring it to this new example code: https://gist.run/?id=4fec87bd40bdfbd7ed10e7ad8f37fdd8 Any thoughts about the throttle tests ?

**Comment by jdanyow** [comment]
re: separate pr for blur:  sounds good

re: throttle- we could refactor `+new Date()` into a function and provide a way for the tests to replace it.

something like:

*throttle-binding-behavior.js*
```js
let getTicks = () => +new Date();

export function setTicksFn(fn) {
  getTicks = fn;
}

...
...
```

**Comment by bigopon** [comment]
It turns out jasmine has `jasmine.clock().mockDate()` to handle that situation But I still cant deal with the throttle test, I think missed something after the main code, been blocking me for 2 days. Can you go ahead and checkout this branch https://github.com/bigopon/templating-resources/tree/templating-resources-debounce to see what I missed in the throttle test ? @jdanyow 

I suppress few tests to deal with failing throttle tests easier.
